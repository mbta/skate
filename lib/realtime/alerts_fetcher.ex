defmodule Realtime.AlertsFetcher do
  @moduledoc false

  use GenServer
  require Logger
  alias Schedule.Route

  @default_poll_interval_ms 60 * 1_000

  @type state :: %{
          update_fn: Route.by_id([String.t()]),
          poll_interval_ms: integer(),
          api_url: String.t(),
          api_key: String.t()
        }

  @spec start_link(Keyword.t()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    {name, opts} = Keyword.pop(opts, :name, __MODULE__)

    GenServer.start_link(__MODULE__, opts, name: name)
  end

  @impl true
  def init(opts) do
    update_fn = Keyword.get(opts, :update_fn, &Realtime.Server.update_alerts/1)

    poll_interval_ms = Keyword.get(opts, :poll_interval_ms, @default_poll_interval_ms)

    {:ok, _} = :timer.send_interval(poll_interval_ms, self(), :query_api)

    {:ok,
     %{
       update_fn: update_fn,
       poll_interval_ms: poll_interval_ms,
       api_url: Application.get_env(:skate, :api_url),
       api_key: Application.get_env(:skate, :api_key)
     }, {:continue, :initial_poll}}
  end

  @impl true
  def handle_continue(:initial_poll, state) do
    do_poll(state)
  end

  @impl true
  def handle_info(:query_api, state) do
    do_poll(state)
  end

  @spec do_poll(state()) :: {:noreply, state()}
  defp do_poll(%{update_fn: update_fn, api_url: api_url, api_key: api_key} = state) do
    headers =
      if api_key do
        [{"x-api-key", api_key}]
      else
        []
      end

    url =
      api_url
      |> URI.merge(
        "/alerts?" <>
          URI.encode_query([{"filter[route_type]", "3,11"}, {"filter[datetime]", "NOW"}])
      )
      |> URI.to_string()

    case HTTPoison.get(url, headers) do
      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
        case parse_and_update_alerts(body, update_fn) do
          :ok ->
            Logger.info("updated_alerts url=#{inspect(url)}")

          {:error, error} ->
            Logger.warning("unable_to_parse_alerts url=#{inspect(url)} error=#{inspect(error)}")
        end

      response ->
        Logger.warning("unexpected_response url=#{inspect(url)} response=#{inspect(response)}")
    end

    {:noreply, state}
  end

  @spec parse_and_update_alerts(binary(), (Route.by_id([String.t()]) -> term())) ::
          :ok | {:error, any()}
  defp parse_and_update_alerts(body, update_fn) do
    case JsonApi.parse(body) do
      %JsonApi{data: data} ->
        alerts =
          Enum.reduce(data, %{}, fn alert, acc ->
            is_detour? = alert.attributes["effect"] == "DETOUR"

            if is_detour? do
              route_ids = route_ids_from_informed_entities(alert.attributes["informed_entity"])

              alert_text = alert.attributes["service_effect"]

              add_alerts_by_routes(route_ids, alert_text, acc)
            else
              acc
            end
          end)

        update_fn.(alerts)

        :ok

      {:error, error} ->
        {:error, error}
    end
  end

  @spec route_ids_from_informed_entities([map()]) :: [Route.id()]
  defp route_ids_from_informed_entities(informed_entities) do
    Enum.reduce(informed_entities, [], fn informed_entity, route_ids ->
      if informed_entity["route"] && informed_entity["route"] not in route_ids do
        [informed_entity["route"] | route_ids]
      else
        route_ids
      end
    end)
  end

  @spec add_alerts_by_routes([Route.id()], String.t(), Route.by_id([String.t()])) ::
          Route.by_id([String.t()])
  defp add_alerts_by_routes(route_ids, alert_text, alerts_by_routes) do
    Enum.reduce(route_ids, alerts_by_routes, fn route_id, acc ->
      {_old_value, acc} =
        Map.get_and_update(acc, route_id, fn alerts_for_route ->
          case alerts_for_route do
            nil -> {alerts_for_route, [alert_text]}
            alerts -> {alerts_for_route, [alert_text | alerts]}
          end
        end)

      acc
    end)
  end
end
