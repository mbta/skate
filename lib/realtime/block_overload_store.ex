defmodule Realtime.BlockOverloadStore do
  use GenServer

  alias Notifications.NotificationServer
  alias Realtime.Vehicle

  def default_name(), do: __MODULE__

  @spec start_link(Keyword.t()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, __MODULE__)
    GenServer.start_link(__MODULE__, nil, name: name)
  end

  @spec update([Vehicle.t()], GenServer.name()) :: :ok
  def update(overloaded_vehicles, server \\ default_name()) do
    GenServer.cast(server, {:update, overloaded_vehicles})
    :ok
  end

  @impl GenServer
  def init(_) do
    {:ok, nil}
  end

  @impl GenServer
  def handle_cast({:update, overloaded_vehicles}, state) do
    new_vehicles =
      if state do
        old_overload_identifiers = Enum.map(state, &overload_identifier/1)

        Enum.reject(
          overloaded_vehicles,
          &Enum.member?(old_overload_identifiers, overload_identifier(&1))
        )
      else
        []
      end

    new_block_overloads_fn =
      Application.get_env(
        :skate,
        :new_block_overloads_fn,
        &NotificationServer.new_block_overloads/1
      )

    _ = new_block_overloads_fn.(new_vehicles)

    {:noreply, overloaded_vehicles}
  end

  defp overload_identifier(vehicle) do
    Map.take(vehicle, [:block_id, :vehicle_id])
  end
end
