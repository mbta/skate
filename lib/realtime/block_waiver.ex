defmodule Realtime.BlockWaiver do
  @moduledoc """
  A representation of known skipped work on a trip, entered by dispatch.
  Essentially this summarizes a set of StopTimeUpdate datapoints.
  """

  alias Concentrate.StopTimeUpdate
  alias Gtfs.{Block, Trip}
  alias Realtime.StopTimeUpdateStore

  @type t :: %__MODULE__{
          start_time: Util.Time.time_of_day(),
          end_time: Util.Time.time_of_day(),
          remark: String.t() | nil
        }

  @enforce_keys [
    :start_time,
    :end_time
  ]

  @derive Jason.Encoder

  defstruct [
    :start_time,
    :end_time,
    :remark
  ]

  @spec block_waivers_for_block(Block.t() | nil) :: [t()]
  def block_waivers_for_block(nil), do: []

  def block_waivers_for_block(block) do
    block
    |> Enum.flat_map(&trip_stop_time_waivers/1)
    |> group_consecutive_sequences()
    |> Enum.map(&from_trip_stop_time_waivers(&1, date_for_block(block)))
  end

  @type trip_stop_time_waiver :: {Util.Time.time_of_day(), StopTimeUpdate.t() | nil}
  @spec trip_stop_time_waivers(Trip.t()) :: [trip_stop_time_waiver()]
  def trip_stop_time_waivers(trip) do
    stop_time_updates_fn =
      Application.get_env(
        :realtime,
        :stop_time_updates_fn,
        &StopTimeUpdateStore.stop_time_updates_for_trip/1
      )

    stop_time_updates = stop_time_updates_fn.(trip.id)

    Enum.map(trip.stop_times, fn stop_time ->
      {stop_time.time,
       Enum.find(stop_time_updates, &(StopTimeUpdate.stop_id(&1) == stop_time.stop_id))}
    end)
  end

  @spec group_consecutive_sequences([trip_stop_time_waiver()], [[trip_stop_time_waiver()]], [
          trip_stop_time_waiver()
        ]) :: [[trip_stop_time_waiver()]]
  def group_consecutive_sequences(trip_stop_time_waivers, acc \\ [], working \\ [])

  def group_consecutive_sequences([], acc, working) do
    bank_working(acc, working)
  end

  def group_consecutive_sequences([{_time, nil} | tail], acc, working) do
    group_consecutive_sequences(tail, bank_working(acc, working))
  end

  def group_consecutive_sequences([trip_stop_time_waiver | tail], acc, []) do
    group_consecutive_sequences(tail, acc, [trip_stop_time_waiver])
  end

  def group_consecutive_sequences(
        [{time, stop_time_update} = trip_stop_time_waiver | tail],
        acc,
        working
      ) do
    {previous_time, previous_stop_time_update} = List.last(working)

    if same_remark?(stop_time_update, previous_stop_time_update) &&
         within_60_minutes?(time, previous_time) do
      group_consecutive_sequences(tail, acc, working ++ [trip_stop_time_waiver])
    else
      # Make a new group if the remark changes or there is more than 60 minutes between stops
      group_consecutive_sequences(tail, bank_working(acc, working), [trip_stop_time_waiver])
    end
  end

  @spec bank_working([[trip_stop_time_waiver()]], [trip_stop_time_waiver()]) :: [
          [trip_stop_time_waiver()]
        ]
  defp bank_working(acc, []), do: acc
  defp bank_working([], working), do: [working]
  defp bank_working(acc, working), do: acc ++ [working]

  @spec same_remark?(map(), map()) :: boolean
  defp same_remark?(%StopTimeUpdate{remark: remark1}, %StopTimeUpdate{remark: remark2}),
    do: remark1 == remark2

  @spec within_60_minutes?(Util.Time.time_of_day(), Util.Time.time_of_day()) :: boolean
  defp within_60_minutes?(time, previous_time), do: time - previous_time <= 3600

  @spec from_trip_stop_time_waivers([trip_stop_time_waiver()], Date.t()) :: t()
  def from_trip_stop_time_waivers(trip_stop_time_waivers, date_for_block) do
    {start_time, %StopTimeUpdate{remark: remark}} = List.first(trip_stop_time_waivers)
    {end_time, _} = List.last(trip_stop_time_waivers)

    %__MODULE__{
      start_time: Util.Time.timestamp_for_time_of_day(start_time, date_for_block),
      end_time: Util.Time.timestamp_for_time_of_day(end_time, date_for_block),
      remark: remark
    }
  end

  @spec date_for_block(Block.t()) :: Date.t()
  defp date_for_block(block) do
    active_blocks_fn = Application.get_env(:realtime, :active_blocks_fn, &Gtfs.active_blocks/2)

    now = Util.Time.now()
    one_hour_ago = now - 60 * 60
    in_ten_minutes = now + 10 * 60

    {date, _blocks} =
      one_hour_ago
      |> active_blocks_fn.(in_ten_minutes)
      |> Enum.find(
        {Util.Time.today(), []},
        fn {_date, blocks} -> Enum.member?(blocks, block) end
      )

    date
  end
end
