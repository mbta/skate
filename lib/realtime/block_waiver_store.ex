defmodule Realtime.BlockWaiverStore do
  @moduledoc """
  Server for setting and getting StopTimeUpdates.
  """

  use GenServer

  alias Schedule.Block
  alias Schedule.Gtfs.Service
  alias Realtime.BlockWaiver

  @type t :: %__MODULE__{
          block_waivers_by_block_key: BlockWaiver.block_waivers_by_block_key(),
          never_set: boolean()
        }

  defstruct block_waivers_by_block_key: %{},
            never_set: true

  # Client

  @spec default_name() :: GenServer.name()
  def default_name(), do: Realtime.BlockWaiverStore

  @spec start_link() :: GenServer.on_start()
  @spec start_link(Keyword.t()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, default_name())

    GenServer.start_link(__MODULE__, %__MODULE__{}, name: name)
  end

  @spec block_waivers_for_block_and_service(Block.id(), Service.id()) :: [BlockWaiver.t()]
  @spec block_waivers_for_block_and_service(Block.id(), Service.id(), GenServer.server()) :: [
          BlockWaiver.t()
        ]
  def block_waivers_for_block_and_service(block_id, service_id, server \\ default_name()) do
    GenServer.call(server, {:block_waivers_for_block_and_service, block_id, service_id})
  end

  @spec set(BlockWaiver.block_waivers_by_block_key()) :: :ok
  @spec set(BlockWaiver.block_waivers_by_block_key(), GenServer.server()) :: :ok
  def set(block_waivers_by_block_key, server \\ default_name()) do
    GenServer.cast(server, {:set, block_waivers_by_block_key})
  end

  # Server

  @impl GenServer
  def init(initial_state) do
    {:ok, initial_state}
  end

  @impl true
  def handle_call(
        {:block_waivers_for_block_and_service, block_id, service_id},
        _from,
        %__MODULE__{
          block_waivers_by_block_key: block_waivers_by_block_key
        } = state
      ) do
    block_waivers = Map.get(block_waivers_by_block_key, {block_id, service_id}, [])

    {:reply, block_waivers, state}
  end

  @impl true
  def handle_cast(
        {:set, new_block_waivers_by_block_key},
        %__MODULE__{} = state
      ) do
    # We have this never_set flag to detect if this is the first time this
    # particular instance of BlockWaiverStore has had set called on it.
    # Consider what happens if we don't do this check: the first time
    # set is called, it calls waiver_diff to compare
    # new_block_waivers_by_block_key against the existing state. But
    # since we just initialized that state to an empty map, all block
    # waivers will appear to be new, and we end up possibly sending
    # notifications for waivers to users who have already seen them.
    #
    if(!state.never_set) do
      new_block_waivers =
        waiver_diff(state.block_waivers_by_block_key, new_block_waivers_by_block_key)

      notification_server_new_block_waivers_fn =
        Application.get_env(
          :notifications,
          :notifications_server_new_block_waivers_fn,
          &Notifications.NotificationServer.new_block_waivers/1
        )

      notification_server_new_block_waivers_fn.(new_block_waivers)
    end

    {:noreply,
     %__MODULE__{
       state
       | block_waivers_by_block_key: new_block_waivers_by_block_key,
         never_set: false
     }}
  end

  @spec waiver_diff(
          BlockWaiver.block_waivers_by_block_key(),
          BlockWaiver.block_waivers_by_block_key()
        ) :: %{Block.key() => MapSet.t()}
  defp waiver_diff(old_block_waivers_by_block_key, new_block_waivers_by_block_key) do
    Enum.reduce(
      new_block_waivers_by_block_key,
      %{},
      fn {block_key, new_block_waiver_list}, result ->
        old_block_waiver_list = Map.get(old_block_waivers_by_block_key, block_key, [])
        old_block_waiver_set = MapSet.new(old_block_waiver_list)

        newly_appearing_waivers =
          new_block_waiver_list
          |> MapSet.new()
          |> MapSet.difference(old_block_waiver_set)

        if MapSet.size(newly_appearing_waivers) == 0 do
          result
        else
          Map.put(result, block_key, MapSet.to_list(newly_appearing_waivers))
        end
      end
    )
  end
end
