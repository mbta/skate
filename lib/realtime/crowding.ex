defmodule Realtime.Crowding do
  @type t :: %__MODULE__{
          load: non_neg_integer() | nil,
          capacity: non_neg_integer() | nil,
          occupancy_status: String.t(),
          occupancy_percentage: float() | nil
        }

  @derive Jason.Encoder
  defstruct [:load, :capacity, :occupancy_status, :occupancy_percentage]
end
