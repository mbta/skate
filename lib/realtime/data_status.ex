defmodule Realtime.DataStatus do
  alias Realtime.Vehicle

  @type t :: :good | :outage

  @spec data_status([Vehicle.t()]) :: t()
  def data_status(vehicles) do
    now_fn = Application.get_env(:skate, :now_fn, &Util.Time.now/0)
    now = now_fn.()

    {good_count, bad_count} =
      vehicles
      |> Enum.filter(&should_consider?/1)
      |> count_partition(fn vehicle -> vehicle_is_good(vehicle, now) end)

    total_count = good_count + bad_count

    if bad_count >= 2 and bad_count / total_count > 0.20 do
      :outage
    else
      :good
    end
  end

  # Our data often has lots of out of service buses that have
  # good reasons for going stale.
  # Only consider buses that we expect and need to have good data
  @spec should_consider?(Vehicle.t()) :: boolean()
  defp should_consider?(vehicle) do
    vehicle.route_id != nil
  end

  @spec vehicle_is_good(Vehicle.t(), Util.Time.timestamp()) :: boolean()
  defp vehicle_is_good(vehicle, now) do
    age_seconds = now - vehicle.timestamp
    # 2 and a half minutes ago.
    age_seconds < 150
  end

  @spec count_partition([element], (element -> boolean())) ::
          {non_neg_integer(), non_neg_integer()}
        when element: term()
  defp count_partition(list, predicate) do
    Enum.reduce(list, {0, 0}, fn element, {prev_trues, prev_falses} ->
      if predicate.(element) do
        {prev_trues + 1, prev_falses}
      else
        {prev_trues, prev_falses + 1}
      end
    end)
  end
end
