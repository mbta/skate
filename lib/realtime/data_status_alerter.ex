defmodule Realtime.DataStatusAlerter do
  use GenServer
  require Logger

  def default_name, do: __MODULE__

  @spec start_link(Keyword.t()) :: GenServer.on_start()
  def start_link(opts \\ []) do
    name = Keyword.get(opts, :name, default_name())

    subscribe_fn =
      Keyword.get(
        opts,
        :subscribe_fn,
        &Realtime.DataStatusPubSub.subscribe/0
      )

    initial_state = %{subscribe_fn: subscribe_fn}

    GenServer.start_link(__MODULE__, initial_state, name: name)
  end

  @impl GenServer
  def init(%{subscribe_fn: subscribe_fn}) do
    _ = subscribe_fn.()
    {:ok, nil}
  end

  @impl GenServer
  def handle_info({:new_data_status, data_status}, state) do
    if data_status == :outage do
      Logger.warn("Data outage detected data_outage_detected")
    end

    {:noreply, state}
  end
end
