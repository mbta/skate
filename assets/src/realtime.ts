import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
