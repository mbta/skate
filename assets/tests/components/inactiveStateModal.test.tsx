import React from "react"
import renderer from "react-test-renderer"
import InactiveNotificationModal from "../../src/components/inactiveNotificationModal"
import { Notification, NotificationState } from "../../src/realtime.d"

describe("InactiveNotificationModal", () => {
  const notification: Notification = {
    id: "123",
    createdAt: new Date(),
    reason: "other",
    routeIds: [],
    runIds: [],
    tripIds: ["123", "456", "789"],
    operatorName: null,
    operatorId: null,
    routeIdAtCreation: null,
    startTime: new Date("2020-10-05"),
    state: "unread" as NotificationState,
  }
  const futureNotification = {
    ...notification,
    startTime: new Date("20200-10-05"),
  }

  test("renders for a notification with no runs", () => {
    const tree = renderer
      .create(<InactiveNotificationModal notification={notification} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders for a notification with one current or past run", () => {
    const tree = renderer
      .create(
        <InactiveNotificationModal
          notification={{ ...notification, runIds: ["111"] }}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders for a notification with multiple current or past runs", () => {
    const tree = renderer
      .create(
        <InactiveNotificationModal
          notification={{ ...notification, runIds: ["111", "222"] }}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders for a notification with one upcoming run", () => {
    const tree = renderer
      .create(
        <InactiveNotificationModal
          notification={{ ...futureNotification, runIds: ["111"] }}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test("renders for a notification with multiple upcoming runs", () => {
    const tree = renderer
      .create(
        <InactiveNotificationModal
          notification={{ ...futureNotification, runIds: ["111", "222"] }}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
