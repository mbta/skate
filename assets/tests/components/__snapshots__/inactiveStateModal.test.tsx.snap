// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InactiveNotificationModal renders for a notification with multiple current or past runs 1`] = `
Array [
  <div
    className="c-modal"
  >
    <div
      className="m-inactive-notification-modal__close-button"
    >
      <button
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-notification__title"
    >
      OTHER
       NOTIFICATION
    </div>
    <div
      className="m-inactive-notification-modal__body"
    >
      Runs 111, 222 are not currently active in Skate. This notification may no longer be relevant.
    </div>
  </div>,
  <div
    className="c-modal-overlay"
  />,
]
`;

exports[`InactiveNotificationModal renders for a notification with multiple upcoming runs 1`] = `
Array [
  <div
    className="c-modal"
  >
    <div
      className="m-inactive-notification-modal__close-button"
    >
      <button
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-notification__title"
    >
      OTHER
       NOTIFICATION
    </div>
    <div
      className="m-inactive-notification-modal__body"
    >
      Runs 111, 222 are upcoming and not yet active in Skate. Please check back later to see details for these runs.
    </div>
  </div>,
  <div
    className="c-modal-overlay"
  />,
]
`;

exports[`InactiveNotificationModal renders for a notification with no runs 1`] = `
Array [
  <div
    className="c-modal"
  >
    <div
      className="m-inactive-notification-modal__close-button"
    >
      <button
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-notification__title"
    >
      OTHER
       NOTIFICATION
    </div>
    <div
      className="m-inactive-notification-modal__body"
    >
      No runs associated with this notification are currently active in Skate. This notification may no longer be relevant.
    </div>
  </div>,
  <div
    className="c-modal-overlay"
  />,
]
`;

exports[`InactiveNotificationModal renders for a notification with one current or past run 1`] = `
Array [
  <div
    className="c-modal"
  >
    <div
      className="m-inactive-notification-modal__close-button"
    >
      <button
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-notification__title"
    >
      OTHER
       NOTIFICATION
    </div>
    <div
      className="m-inactive-notification-modal__body"
    >
      Run 111 is not currently active in Skate. This notification may no longer be relevant.
    </div>
  </div>,
  <div
    className="c-modal-overlay"
  />,
]
`;

exports[`InactiveNotificationModal renders for a notification with one upcoming run 1`] = `
Array [
  <div
    className="c-modal"
  >
    <div
      className="m-inactive-notification-modal__close-button"
    >
      <button
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </div>
    <div
      className="m-notification__title"
    >
      OTHER
       NOTIFICATION
    </div>
    <div
      className="m-inactive-notification-modal__body"
    >
      Run 111 is upcoming and not yet active in Skate. Please check back later to see details for this run.
    </div>
  </div>,
  <div
    className="c-modal-overlay"
  />,
]
`;
